### PR TITLE
feature/SOF-7273

### PR DESCRIPTION
**Pull request by VsevolodX**
This PR makes it possible to develop and debug notebooks for Pyodide JL in Python JL environment with minimal adjustments to the code (we usually have to comment out installation anyways).

## Environment detection: 
utils.py detect if this is Pyodide environment, and such global flag is used by other functions to switch behavior. 
## Installation: 
`micropip` commands are used in Pyodide, while `pip` in Python.
## Read (materials) data: 
in Pyodide the message to JS extension is sent, while in Python data is read from `uploads` folder.
## Send (materials) data: 
in Pyodide again the message to JS, while in Python data is saved to a file in `uploads` folder.

### CODE REVIEW

File: other/jupyterlite/utils.py [hunk 1]
```diff
@@ -2,63 +2,96 @@
 import json
 import os
 
+UPLOADS_FOLDER = "uploads"
+
+# Environment detection
+IN_PYODIDE = False
```
Comment (VsevolodX) [inline other/jupyterlite/utils.py:8]: IS_RUNNING_IN_PYODIDE

File: other/jupyterlite/utils.py [hunk 2]
```diff
@@ -71,35 +104,64 @@ def set_data(key, value):
...
+            }} else {{
+                console.error('requestDataFromHost function is not defined on the window object.')
+            }}
+        }})();
+        """
+        display(Javascript(js_code))
+        print(f"Status: {key} requested")
+    else:
+        # JupyterLab environment: Read data from the 'uploads' folder
+        try:
```
Comment (VsevolodX) [inline other/jupyterlite/utils.py:155]: Move into separate functions

File: other/materials_designer/config.yml [hunk 3]
```diff
@@ -1,4 +1,9 @@
 # This manifest file is used to install the required packages for the corresponding notebook.
+default:
+  packages:
+    - lzma
+    - sqlite3
+notebooks:
 - notebook: create_interface_with_min_strain_zsl.ipynb
```
Comment (VsevolodX) [inline other/materials_designer/config.yml:7]: name

File: other/materials_designer/config.yml [hunk 4]
```diff
@@ -1,4 +1,9 @@
 # This manifest file is used to install the required packages for the corresponding notebook.
+default:
+  packages:
```
Comment (VsevolodX) [inline other/materials_designer/config.yml:3]:
```yaml
# This manifest file is used to install the required packages for the corresponding notebook.
# The following types of kernels are supported:
# 1. python
# 2. pyodide
# Accordingly, the packages are installed using the kernel type, e.g., packages_pyodide.
# The default packages are installed for all the notebooks.
default:
  packages_common:
  packages_python:
  packages_pyodide:
    - lzma
    - sqlite3
notebooks:
- notebook: create_interface_with_min_strain_zsl.ipynb
  packages_common:
    - https://files.mat3ra.com:44318/uploads/pymatgen-2023.9.10-py3-none-any.whl
    - https://files.mat3ra.com:44318/web/pyodide/spglib-2.0.2-py3-none-any.whl
    - https://files.pythonhosted.org/packages/d9/0e/2a05efa11ea33513fbdf4a2e2576fe94fd8fa5ad226dbb9c660886390974/ruamel.yaml-0.17.32-py3-none-any.whl
    - ase==3.22.1
    - networkx==3.2.1
    - monty==2023.11.3
    - scipy==1.11.2
    - tabulate==0.9.0
    - sympy==1.12
    - uncertainties==3.1.6
    - ipywidgets
    - plotly==5.18
    - pandas==1.5.3
    - nbformat>=4.2.0

- notebook: import_material_from_jarvis_db_entry.ipynb
  packages:
    - https://files.mat3ra.com:44318/uploads/pymatgen-2023.9.10-py3-none-any.whl
    - https://files.mat3ra.com:44318/web/pyodide/spglib-2.0.2-py3-none-any.whl
    - https://files.pythonhosted.org/packages/d9/0e/2a05efa11ea33513fbdf4a2e2576fe94fd8fa5ad226dbb9c660886390974/ruamel.yaml-0.17.32-py3-none-any.whl
    - https://files.mat3ra.com:44318/uploads/paginate-0.5.6-py3-none-any.whl
    - https://files.mat3ra.com:44318/uploads/watchdog-2.3.1-py3-none-any.whl
    - networkx==3.2.1
    - monty==2023.11.3
    - scipy==1.11.2
    - tabulate==0.9.0
    - sympy==1.12
    - uncertainties==3.1.6
    - pandas==1.5.3
    - express-py==2024.2.2.post2
    - mat3ra-esse
    - jarvis-tools
    - munch
    - setuptools

- notebook: import_materials_from_files.ipynb
  packages:
    - $ref: "#/definitions/pymatgen_packages"
    - ase==3.22.1
    - networkx==3.2.1
    - monty==2023.11.3
    - scipy==1.11.2
    - tabulate==0.9.0
    - sympy==1.12
    - uncertainties==3.1.6
    - pandas==1.5.3
    - express-py==2024.2.2.post2
    - mat3ra-esse
    - jarvis-tools
    - munch
    - setuptools
definitions:
  pymatgen_packages:
    - https://files.mat3ra.com:44318/uploads/pymatgen-2023.9.10-py3-none-any.whl
    - https://files.mat3ra.com:44318/web/pyodide/spglib-2.0.2-py3-none-any.whl
    - https://files.pythonhosted.org/packages/d9/0e/2a05efa11ea33513fbdf4a2e2576fe94fd8fa5ad226dbb9c660886390974/ruamel.yaml-0.17.32-py3-none-any.whl
    - https://files.mat3ra.com:44318/uploads/paginate-0.5.6-py3-none-any.whl
    - https://files.mat3ra.com:44318/uploads/watchdog-2.3.1-py3-none-any.whl

```

File: other/jupyterlite/utils.py [hunk 5]
```diff
@@ -73,33 +103,97 @@ def set_data(key, value):
...
+    """
+    try:
+        materials = []
+        for filename in os.listdir(UPLOADS_FOLDER):
+            if filename.endswith(".json"):
+                with open(os.path.join(UPLOADS_FOLDER, filename), "r") as file:
+                    data = json.load(file)
+                name = os.path.splitext(filename)[0]
+                print(f"Data from {name} has been read successfully.")
+                materials.append(data)
```
Comment (VsevolodX) [inline other/jupyterlite/utils.py:182]: items instead of materials

File: other/jupyterlite/utils.py [hunk 6]
```diff
@@ -1,70 +1,100 @@
...
-
-    if packages is None:
-        raise ValueError(f"No packages found for notebook {notebook_name}")  
 
+    # Hash the requirements to avoid re-installing packages
+    requirements_hash = str(hash(json.dumps(requirements)))
     if os.environ.get("requirements_hash") != requirements_hash:
-        for package in packages:
-            await install_package(package)
+        packages = []
```
Comment (VsevolodX) [inline other/jupyterlite/utils.py:71]: packages_default = 
packages_common = 
packages_specific =
packages = [...packages_default, .....]

**Review by timurbazhirov — APPROVED** [review]

**Pull request merged**